OpenACC port of halo exchanges

abishekg7 opened the pull request:
This PR enables execution of halo exchanges on GPUs via OpenACC directives, when the MPAS atmosphere core has been built with an appropriate GPU-aware MPI library. This PR builds on the OpenACC memory consolidation introduced in https://github.com/MPAS-Dev/MPAS-Model/pull/1315. 

Module `mpas_halo` is modified in the following ways to enable GPU-aware halo exchanges:
- In the call to `mpas_halo_exch_group_complete`, OpenACC directives copy to device all the relevant fields and metadata that are required for the packing and unpacking loops later.
- OpenACC directives are introduced around the packing and unpacking loops to perform the field to/from send/recv buffer operations on the device. The `attach` clauses introduced to the parallel constructs ensures that the device pointers are attached to the device targets at the start of the parallel region and detached at the end of the region.
- The actual `MPI_Isend` and `MPI_Irecv` operations use GPU-aware MPI, by wrapping these calls within
```
!$acc host_data use_device(pointer_to_buffer)

!$acc end host_data
```

In addition, this PR introduces a new namelist option under the `development` group, `config_gpu_aware_mpi`. When set to `true`, it switches on GPU-direct halo exchanges. The default setting is `false`, in which halo exchanges occur on the host, which necessitates host<->device data transfers around the halo exchanges as necessary.

Note: This feature is not available to use when config_halo_exch_method is set to 'mpas_dmpar'

mgduda commented on src/framework/mpas_halo.F:
```diff
@@ -508,6 +547,7 @@ subroutine mpas_halo_exch_group_full_halo_exch(domain, groupName, iErr)
         use mpas_derived_types, only : domain_type, mpas_halo_group, MPAS_HALO_REAL, MPAS_LOG_CRIT
         use mpas_pool_routines, only : mpas_pool_get_array
         use mpas_log, only : mpas_log_write
+        use mpas_timer, only : mpas_timer_start, mpas_timer_stop
```
Don't forget to update the `Makefile` to add a dependency on `mpas_timer.o` for the `mpas_halo.o` target.

Commit:
Enable GPU-aware halo exchanges via OpenACC directives

This commit enables execution of halo exchanges on GPUs via OpenACC directives,
if the MPAS atmosphere core has been built with an appropriate GPU-aware MPI
distribution.

Module mpas_halo is modified in the following ways to enable GPU-aware halo
exchanges:

- In the call to mpas_halo_exch_group_complete, OpenACC directives copy to
  device all the relevant fields and metadata that are required for the
  packing and unpacking loops later.
- OpenACC directives are introduced around the packing and unpacking loops to
  perform the field to/from send/recv buffer operations on the device. The
  attach clauses introduced to the parallel constructs ensures that the device
  pointers are attached to the device targets at the start of the parallel
  region and detached at the end of the region.
- The actual MPI_Isend and MPI_Irecv operations use GPU-aware MPI, by wrapping
  these calls within !$acc host_data constructs.

Note: This commit introduces temporary host-device data movements in the
atm_core_init routine around the two calls to exchange_halo_group. This is
required just for this commit as all halo-exchanges occur on the device and
fields not yet present on the device must be copied over to it before the halo
exchanges and back to host after it. These copies will be removed in subsequent
commits.

abishekg7 committed:
New namelist option to switch on or off GPU-aware MPI halo exchanges

Introducing a new namelist option under development, config_gpu_aware_mpi,
which will control whether the MPAS runs on GPUs will use GPU-aware MPI or
perform a device<->host update of variables around the call to a purely
CPU-based halo exchange.

Note: This feature is not available to use when config_halo_exch_method is set
to 'mpas_dmpar'

mgduda commented on src/core_atmosphere/Registry.xml:
```diff
@@ -443,6 +443,10 @@
                      units="-"
                      description="Method to use for exchanging halos"
                      possible_values="`mpas_dmpar', `mpas_halo'"/>
+               <nml_option name="config_gpu_aware_mpi" type="logical" default_value="false"
```
For consistency and readability, let's consider adding a blank line above this. (I do see that the `IAU` namelist record doesn't use a blank line between options, but we can address that in a separate clean-up PR at a later time.)

mgduda commented on src/core_atmosphere/Registry.xml:
```diff
@@ -443,6 +443,10 @@
                      possible_values="`mpas_dmpar', `mpas_halo'"/>
+               <nml_option name="config_gpu_aware_mpi" type="logical" default_value="false"
+                     units="-"
+                     description="Whether to use GPU-aware MPI for halo exchanges"
```
In the description, we could also mention that this only works when `config_halo_exch_method` is set to `mpas_halo`.

mgduda commented on src/framework/mpas_dmpar.F:
```diff
@@ -7450,19 +7450,28 @@ end subroutine mpas_dmpar_exch_group_end_halo_exch!}}}
          iErr = MPAS_DMPAR_NOERR
       end if
 
+      useGPUAwareMPI = .false.
```
Is this local variable actually needed?

mgduda commented on src/core_atmosphere/mpas_atm_halos.F:
```diff
@@ -48,18 +48,23 @@ subroutine atm_build_halo_groups(domain, ierr)
          call mpas_log_write('')
 
+         if (config_gpu_aware_mpi) then
+            call mpas_log_write('GPU-aware MPI is not presently supported with config_halo_exch_method = mpas_dmpar',MPAS_LOG_CRIT)
```
This seems like it's redundant given the check that exists in the `mpas_dmpar_exch_group_full_halo_exch` routine.

abishekg7 replied:
I will retain this check and remove the redundant one in `mpas_dmpar_exch_group_full_halo_exch` so that it will terminate earlier.

mgduda replied:
Actually, I think the check in `mpas_dmpar_exch_group_full_halo_exch` is preferable, since it would catch uses of the routine with `withGPUAwareMPI = .true.` in cores other than the `atmosphere` core.

mgduda commented on src/framework/mpas_halo.F:
```diff
@@ -555,11 +596,15 @@ subroutine mpas_halo_exch_group_full_halo_exch(domain, groupName, iErr)
         integer, dimension(:,:,:), CONTIGUOUS pointer :: recvListSrc, recvListDst
         integer, dimension(:), CONTIGUOUS pointer :: unpackOffsets
 
-
```
I'd suggest leaving this blank line in place, as this change isn't necessary to the objective of this PR and this line wouldn't otherwise be modified.

mgduda commented on src/framework/mpas_halo.F:
```diff
@@ -577,6 +622,12 @@ subroutine mpas_halo_exch_group_full_halo_exch(domain, groupName, iErr)
+            return
+        end if
+
+        call mpas_timer_start('full_halo_exch')
```
In the past, my experience has been that timing for halo exchanges may not be especially valuable. Load imbalance can show up as time apparently spent in halo communications, when it's really just that one MPI task had to finish some extra work before participating in the halo exchange. What do you think about removing these timers?

abishekg7 replied:
That's true. I'm fine with removing it for now and revisiting it as needed.

mgduda commented on src/framework/mpas_halo.F:
```diff
@@ -589,6 +640,7 @@ subroutine mpas_halo_exch_group_full_halo_exch(domain, groupName, iErr)
 #endif
         rank = group % fields(1) % compactHaloInfo(8)
 
+        !$acc data present(group % recvBuf(:), group % sendBuf(:)) if(useGPUAwareMPI)
```
If the `parallel` directives in this routine all use `default(present)`, is this directive necessary?

abishekg7 replied:
Doesn't appear to be required, and will remove it.

mgduda commented on src/core_atmosphere/mpas_atm_halos.F:
```diff
@@ -48,12 +48,14 @@ subroutine atm_build_halo_groups(domain, ierr)
       ! Determine from the namelist option config_halo_exch_method which halo exchange method to employ
       !
       call mpas_pool_get_config(domain % blocklist % configs, 'config_halo_exch_method', config_halo_exch_method)
+      call mpas_pool_get_config(domain % blocklist % configs, 'config_gpu_aware_mpi', config_gpu_aware_mpi)
```
It doesn't look like `config_gpu_aware_mpi` is subsequently used in this routine. Should we just remove it altogether?

abishekg7 replied:
Thanks! Fixed.
